src/persist-client/src/iter.rs

```diff
@@ -0,0 +1,700 @@
+        desc: &Description<T>,
+        parts: impl IntoIterator<Item = &'a HollowBatchPart>,
+    ) {
+        let run = parts
+            .into_iter()
+            .map(|part: &HollowBatchPart| {
+                let c_part = ConsolidationPart::Queued {
+                    data: Some(FetchData {
+                        shard_id,
+                        blob: Arc::clone(blob),
+                        metrics: Arc::clone(metrics),
+                        read_metrics,
+                        shard_metrics: Arc::clone(shard_metrics),
+                        part_key: part.key.clone(),
+                        part_desc: desc.clone(),
+                    }),
+                };
+                let size = part.encoded_size_bytes;
+                (c_part, size)
+            })
+            .collect();
+        self.runs.push(run);
+    }
+
+    /// Tidy up: discard any empty parts, and discard any runs that have no parts left.
+    fn trim(&mut self) {
+        self.runs.retain_mut(|run| {
+            while run.front().map_or(false, |(part, _)| part.is_empty()) {
+                run.pop_front();
+            }
+            !run.is_empty()
+        });
+
+        // Some budget may have just been freed up: start prefetching.
+        self.start_prefetches();
+    }
+
+    fn iter(&mut self) -> Option<ConsolidatingIter<T, D>> {
+        self.trim();
+
+        self.initial_state = self.drop_stash.take();
+        if self.initial_state.is_none() && self.runs.is_empty() {
+            return None;
+        }
+
+        let mut iter = ConsolidatingIter::new(
+            self.initial_state.as_ref().map(borrow_tuple),
+            &mut self.drop_stash,
+        );
+
+        for run in &mut self.runs {
+            let last_in_run = run.len() < 2;
+            if let Some((part, _)) = run.front_mut() {
+                let part_iter = match part {
+                    ConsolidationPart::Encoded { part, cursor } => {
+                        ConsolidationPartIter::encoded(part, cursor, &self.as_of)
+                    }
+                    ConsolidationPart::Sorted { data, index } => {
+                        ConsolidationPartIter::Sorted { data, index }
+                    }
+                    ConsolidationPart::Queued { .. } | ConsolidationPart::Prefetched { .. } => {
+                        // Unreachable from the public API: we always join on the first part of every
+                        // run in `next`, so we should never encounter an unfetched part here.
+                        // TODO: when we have lower bounds on keys in the stats, we could insert a
+                        // placeholder here instead.
+                        panic!("trying to create an interator from an unfetched part!")
+                    }
+                };
+                iter.push(part_iter, last_in_run);
+            }
+        }
+
+        Some(iter)
+    }
+
+    pub(crate) async fn next(&mut self) -> Option<ConsolidatingIter<T, D>> {
```

Comment:
on initial read, I found this signature surprising and wasn't quite sure what it meant until much more diving in. I'm imagining the references make implementing an iterator over `TupleRef<'a, T, D>` directly challenging, so doc comment here would be helpful